src/series/candlestick.js

```diff
@@ -1,62 +1,47 @@
 import d3 from 'd3';
 import dataJoinUtil from '../util/dataJoin';
-import fractionalBarWidth from '../util/fractionalBarWidth';
 import {noop} from '../util/fn';
 import candlestickSvg from '../svg/candlestick';
+import ohlcBase from './ohlcBase';
+import {rebindAll} from '../util/rebind';
 
 export default function() {
 
     var decorate = noop,
-        xScale = d3.time.scale(),
-        yScale = d3.scale.linear(),
-        xValue = function(d, i) { return d.date; },
-        yOpenValue = function(d, i) { return d.open; },
-        yHighValue = function(d, i) { return d.high; },
-        yLowValue = function(d, i) { return d.low; },
-        yCloseValue = function(d, i) { return d.close; },
-        barWidth = fractionalBarWidth(0.75);
+        base = ohlcBase();
 
     var dataJoin = dataJoinUtil()
         .selector('g.candlestick')
         .element('g')
         .attr('class', 'candlestick');
 
-    var xValueScaled = function(d, i) { return xScale(xValue(d, i)); };
-
     var candlestick = function(selection) {
 
         selection.each(function(data, index) {
 
-            var g = dataJoin(this, data);
+            var filteredData = data.filter(base.defined);
+
+            var g = dataJoin(this, filteredData);
 
             g.enter()
                 .append('path');
 
             var pathGenerator = candlestickSvg()
-                    .width(barWidth(data.map(xValueScaled)));
+                    .width(base.computeWidth(filteredData));
 
             g.each(function(d, i) {
 
-                var yCloseRaw = yCloseValue(d, i),
-                    yOpenRaw = yOpenValue(d, i),
-                    x = xValueScaled(d, i),
-                    yOpen = yScale(yOpenRaw),
-                    yHigh = yScale(yHighValue(d, i)),
-                    yLow = yScale(yLowValue(d, i)),
-                    yClose = yScale(yCloseRaw);
+                var values = base.computeValues(d, i);
 
                 var g = d3.select(this)
-                    .classed({
-                        'up': yCloseRaw > yOpenRaw,
-                        'down': yCloseRaw < yOpenRaw
-                    })
-                    .attr('transform', 'translate(' + x + ', ' + yHigh + ')');
+                    .classed(values.classed)
```

Comment:
I'd be tempted to expose `up`/`down` boolean properties rather than wrap up the class name. Mainly because I'm aware of `classed` being a performance nightmare.